crates/provider/src/layers/batch.rs

```diff
@@ -0,0 +1,515 @@
+            target: m3a,
+            allowFailure: true,
+            callData: data.into(),
+        };
+        match self {
+            Self::Call(tx) => IMulticall3::Call3 {
+                target: tx.to().unwrap_or_default(),
+                allowFailure: true,
+                callData: tx.input().cloned().unwrap_or_default(),
+            },
+            Self::BlockNumber => m3a_call(IMulticall3::getBlockNumberCall {}.abi_encode()),
+            Self::ChainId => m3a_call(IMulticall3::getChainIdCall {}.abi_encode()),
+            Self::Balance(addr) => m3a_call(IMulticall3::getEthBalanceCall { addr }.abi_encode()),
+        }
+    }
+}
+
+/// A provider that batches multiple requests into a single request.
+///
+/// See [`BatchLayer`] for more information.
+pub struct BatchProvider<P, N: Network = Ethereum> {
+    provider: Arc<P>,
+    inner: BatchProviderInner,
+    _pd: PhantomData<N>,
+}
+
+impl<P, N: Network> Clone for BatchProvider<P, N> {
+    fn clone(&self) -> Self {
+        Self { provider: self.provider.clone(), inner: self.inner.clone(), _pd: PhantomData }
+    }
+}
+
+impl<P: fmt::Debug, N: Network> fmt::Debug for BatchProvider<P, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("BatchProvider(")?;
+        self.provider.fmt(f)?;
+        f.write_str(")")
+    }
+}
+
+impl<P: Provider<N> + 'static, N: Network> BatchProvider<P, N> {
+    fn new(inner: P, layer: &BatchLayer) -> Self {
+        let inner = Arc::new(inner);
+        let tx = BatchProviderBackend::spawn(inner.clone(), layer);
+        Self { provider: inner, inner: BatchProviderInner { tx, m3a: layer.m3a }, _pd: PhantomData }
+    }
+}
+
+#[derive(Clone)]
+struct BatchProviderInner {
+    tx: mpsc::UnboundedSender<BatchProviderMessage>,
+    m3a: Address,
+}
+
+impl BatchProviderInner {
+    /// We only want to perform a scheduled multicall if:
+    /// - The request has no block ID or state overrides,
+    /// - The request has a target address,
+    /// - The request has no other properties (`nonce`, `gas`, etc cannot be sent with a multicall).
+    ///
+    /// Ref: https://github.com/wevm/viem/blob/ba8319f71503af8033fd3c77cfb64c7eb235c6a9/src/actions/public/call.ts#L295
+    fn should_batch_call<N: Network>(&self, params: &crate::EthCallParams<N>) -> bool {
+        // TODO: block ID is not yet implemented
+        if params.block().is_some_and(|block| block != BlockId::latest()) {
+            return false;
+        }
+        if params.overrides.as_ref().is_some_and(|overrides| !overrides.is_empty()) {
+            return false;
+        }
+        let tx = params.data();
+        if tx.to().is_none() {
+            return false;
+        }
+        macro_rules! any_is_some {
+            ($($name:ident),* $(,)?) => { false $(|| tx.$name().is_some())* };
+        }
+        // TODO: Not exhaustive, should we serialize to JSON and check that instead?
+        if any_is_some!(
```

Comment:
I think this is also missing the blob fee param?

Reply:
I copy pasted every method that is available in the trait